crates/astria-sequencer/src/service/info.rs

```diff
@@ -72,126 +159,119 @@ impl Service<InfoRequest> for Info {
     fn call(&mut self, req: InfoRequest) -> Self::Future {
         let span = req.create_span();
 
-        handle_info_request(self.storage.clone(), req)
+        self.clone()
+            .handle_info_request(req)
             .instrument(span)
             .boxed()
     }
 }
 
-#[instrument(skip(storage))]
-async fn handle_info_request(
-    storage: Storage,
-    request: InfoRequest,
-) -> Result<InfoResponse, BoxError> {
-    match &request {
-        InfoRequest::Info(_) => {
-            let response = InfoResponse::Info(response::Info {
-                version: "0.1.0".to_string(),
-                app_version: 1,
-                last_block_height: Height::default(),
-                last_block_app_hash: AppHash::default(),
-                data: "astria_sequencer".to_string(),
-            });
-            Ok(response)
-        }
-        InfoRequest::Echo(echo) => Ok(InfoResponse::Echo(Echo {
-            message: echo.message.clone(),
-        })),
-        InfoRequest::Query(req) => Ok(InfoResponse::Query(
-            handle_query(storage, req)
-                .await
-                .context("failed handling query request")?,
-        )),
-        // this was removed after v0.34
-        InfoRequest::SetOption(_) => Ok(InfoResponse::SetOption(SetOption {
-            code: Code::default(),
-            log: "SetOption is not supported".to_string(),
-            info: "SetOption is not supported".to_string(),
-        })),
+struct BoxedAbciQueryHandler(Box<dyn ErasedAbciQueryHandler>);
```

Comment:
what is this type for (and ErasedAbciQueryHandler?)

Reply:
Yes, fundamentally that is the case.

I am going to write down my reasoning as to why we ended up with these sevaral layers of complexity:

1. I want to register one handler per matched path, so that we can quickly associate `path -> handler`;
2. first problem: every handler is a `async fn -> ret`, which is sugar for `fn -> impl Fut<Out=ret>`. So because every handler returns an anonymous type, each handler itself is an anonymous type;
3. there are 2 ways around this:
    a. change the function signature to `fn -> Box<dyn Fut<Out=ret>>`
    b. box the entire function, i.e. have a a `Box<dyn fn -> impl Fut=ret>`
4. I chose option b. because that way the function signature of the handler does not have to change; `trait AbciQueryHandler` just makes this all nicer because I can avoid wirting all the `Clone + Send + etc` stuff line.
5. Now, `struct Info` is `Clone`, so `matchit::Router<T>` must be clone, but `AbciQueryHandler: Clone`, so it's not typesafe, meaning, I can't make a `Box<dyn AbciQueryHandler>` from it.
6. This means we need to introduce another `trait ErasedAbciQueryHandler` for it that is not clone. But you see it has `ErasedAbciQueryHandler::clone_box`, so I can implement `Clone for BoxedAbciQueryHandler` in terms of it.
7.  Finally, I have to have a way to go from a handler (that implements `AbciQueryHandler`) to something that implements `ErasedAbciQueryAndler`. This is where `MakeErasedAbciQueryHandler` comes in as a wrapper around the handler.

Reply:
wow a lot of steps, thanks for explaining!

Reply:
@joroshiba I have moved the router impl to a submodule under `service::info` and explained the implementation there.